Add `FinalizerSafe` auto trait to enforce safe drop impls for `Gc` values

jacob-hughes opened the pull request:
This change lets us accept *a lot* more types for `Gc<T>`. In f1757e, we check if `T` implements `Drop`, and if it does, we then check that `T: Send + Sync`. This commit improves the algorithm to now do the following:

1. Instead, check if `T: FinalizerSafe`, which is set up to have all the same bounds as `T: Send + Sync` in the standard library.

2. If `T: !FinalizerSafe`, look at all the drop methods that `T` needs to call. That means `T`s drop method, and any drop methods of `T`s component types.

3. Look at the MIR for each of these drop methods, if any of them contain a projection into a field which is not `FinalizerSafe` then emit a compiler error, showing the exact line inside `drop` which is unsound.

This makes `Gc` much safer to use, because there are types which implement `Drop` which are not `Send + Sync`, but they don't do anything in `Drop` in a thread-unsafe way. We can now make use of Rust's ability to detect races at compile-time to ensure we have safe drop calls inside finalizers.

This also lets us prevent users from derefing `Gc` fields inside finalizers, as we mark `Gc` as `!FinalizerSafe` *but* leave it `Send + Sync` if its contents are.

This can be better understood with an example. Consider the following Rust program which performs a non-thread-safe operation inside its `drop` method.

```rust

    struct S(Rc<Cell<String>>);

    impl !std::marker::FinalizerSafe for S{}

    impl Drop for S {
        fn drop(&mut self) {
            self.0.replace("World".to_string());
        }
    }

    fn main() {
        let rc = Rc::new(Cell::new("Hello".to_string()));
        let gc = Gc::new(S(Rc::clone(&rc)));
    }

```
This will give the following compiler error:

```
error: `S(Rc::clone(&rc))` cannot be safely finalized.
  --> src/main.rs:19:22
   |
13 |         self.0.replace("World".to_string());
   |         -----------------------------------
   |         |
   |         caused by the expression in `fn drop(&mut)` here because
   |         it uses a type which is not safe to use in a finalizer.
...
19 |     let gc = Gc::new(S(Rc::clone(&rc)));
   |                      ^^^^^^^^^^^^^^^^^ has a drop method which cannot be safely finalized.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.
```

Here's another example where we erroneously use a `Gc` field inside a finalizer:

```rust
struct HasGcFields(Gc<usize>);

impl Drop for HasGcFields {
    fn drop(&mut self) {
        println!("Boom {}", self.0);
    }
}

fn main() {
    Gc::new(HasGcFields(Gc::new(123)));
}
```

Which gives the following error:

```
error: `HasGcFields(Gc::new(123))` cannot be safely finalized.
  --> src/main.rs:18:13
   |
13 |         println!("Boom {}", self.0);
   |                             ------
   |                             |
   |                             caused by the expression in `fn drop(&mut)` here because
   |                             it uses another `Gc` type.
...
18 |     Gc::new(HasGcFields(Gc::new(123)));
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^ has a drop method which cannot be safely finalized.
   |
   = help: `Gc` finalizers are unordered, so this field may have already been dropped. It is not safe to dereference.
```

ltratt commented on compiler/rustc_mir_transform/src/check_finalizers.rs:
```diff
@@ -70,18 +48,186 @@ impl<'tcx> MirPass<'tcx> for CheckFinalizers {
+        for (_, proj) in place_ref.iter_projections() {
+            match proj {
+                ProjectionElem::Field(_, ty) => {
+                    // if (!self.cx.is_send(ty) && !self.cx.is_sync(ty))
```
Commented code.

jacob-hughes replied:
Oops! Fixed here: 32e8f65183c04286001f9379f08b4c200953e513

ltratt commented on library/core/src/marker.rs:
```diff
@@ -482,6 +482,28 @@ impl<T: ?Sized> !Sync for *const T {}
 
+#[unstable(feature = "gc", issue = "none")]
+#[cfg_attr(not(test), rustc_diagnostic_item = "FinalizerSafe")]
+/// Types for which it is safe to use inside a finalizer.
```
Maybe this first sentence should be something like  "A marker trait used to state that a struct's drop method can safely be used with GC finalisation." ?

ltratt commented on library/core/src/marker.rs:
```diff
@@ -482,6 +482,28 @@ impl<T: ?Sized> !Sync for *const T {}
+/// cases, this is already covered if every type used inside the drop method
+/// implements `Send` + `Sync`.
+///
+/// The `FinalizerSafe` trait means that a type is safe to use inside a `Drop`
```
I can't quite parse this and I think it doesn't quite covere everything. I *think* it wants to say something like "This marker trait can be explicitly used to inform the type system that a struct's `drop` method is `Send+Sync`-safe even if the struct as a whole is not. For example, if a struct `S` has a single field `f` whose type is not `Send+Sync` but `S`'s `drop` method does not reference that field, then the struct as a whole can be safely marked `FinalizerSafe`. Note that this trait only affects GC finalisation and has no effect outside the `Gc` type."

jacob-hughes replied:
Yeah that's right. I can change the comment to look more like this. I guess the extra bit though which I'm not sure how to add without maybe confusing things is the following.

It's also valid to call `Gc<T>::new` on a `T: !FinalizerSafe`, provided that `T`'s drop method never references one of it's fields which is `!FinalizerSafe`.

The cool part of this PR is that now we check this statically, whereas before I had to manually check every `Drop` method in WLambda was safe.

ltratt replied:
> It's also valid to call Gc<T>::new on a T: !FinalizerSafe, provided that T's drop method never references one of its fields which is !FinalizerSafe.

That doesn't quite make sense to me, but maybe I've misunderstood. If `T` is not `Send+Sync` and not `FinalizerSafe` surely `Gc::new` will just reject it?

In terms of the comment I might tweak it slightly to:

> This marker trait can be explicitly used to inform the type system that a struct's drop method is Send+Sync-safe even if the struct as a whole is not. For example, if a struct S has a single field f whose type is not Send+Sync but S's drop method does not reference that field, then the struct as a whole can be safely marked `FinalizerSafe`.
>
> Note that this trait only has relevance to GC finalisation and `Gc::new`. In other words, `FinalizerSafe` does not affect a struct when it is not embedded in a `Gc`.

jacob-hughes replied:
> That doesn't quite make sense to me, but maybe I've misunderstood. If T is not Send+Sync and not FinalizerSafe surely Gc::new will just reject it?

It will reject it only if it can't see into its drop method. For example, the following program would be accepted, even though `A: !FinalizerSafe`:

```rust

// A is !Sync and !FinalizerSafe because it contains a `Cell`. 
struct A {
   field1: usize,
   field2: Cell<usize>
}

impl Drop for A {
  fn drop(&mut self) {
    println!("Hello world")
  }
}

fn main() {
  let a = A{...};
  Gc::new(a);
}
```

This is fine because `A::Drop` doesn't dereference anything that is `!FinalizerSafe`. However, if we changed `A`'s drop method to the following:

```rust
impl Drop for A {
  fn drop(&mut self) {
    println!("Hello {}", self.field2.get())
  }
}

```

Now we'd get the following compiler error:

```

error: `a` cannot be safely finalized.
  --> src/main.rs:27:13
   |
18 |         println!("Hello {:?}", self.field2.get())
   |                                -----------------
   |                                |
   |                                caused by the expression in `fn drop(&mut)` here because
   |                                it uses a type which is not safe to use in a finalizer.
...
27 |     Gc::new(a);
   |             ^ has a drop method which cannot be safely finalized.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `Send + Sync` or `FinalizerSafe`.
```

However, if `A` was a type that we couldn't see into at compile time (e.g. it's a trait object, or it hasn't been fully monomorphized), then we would have to reject `A` when it's passed into `Gc::new` because it's `!FinalizerSafe` and we don't know what its drop method does.

ltratt replied:
Oh, we really *do* look inside the `drop` method? Is that recursive? e.g. if in a `drop` method I call `f(&mut self)` will `f` be checked to see if it does anything naughty?

jacob-hughes replied:
>Oh, we really do look inside the drop method? 

Yes that's right!

>Is that recursive?

It's not recursive, no. Right now if we call a function that takes a field that it shouldn't touch as an argument then we get a compiler error. Your example is a good one because it takes a `&mut self`, this should also fail because we don't know what `f` will do with `&mut self`. You've discovered a bug, because we should reject this (as we already do if we had a function `g(&mut Cell<usize>)` and passed in `self.field2`. I shall fix this!

jacob-hughes replied:
Even though it's not recursive though, it's already surprisingly effective (on standard library types and WLambda at least) because we attach the pass after inlining and most `Drop` implementations don't really do anything too complicated other than perhaps a decref / dealloc on particular fields.

jacob-hughes replied:
To clarify though: where we can't recurse through a function call to know that it's safe, we reject the program at compile time. So all we're doing here is easing up a bit and letting more (but obviously not all) correct programs through. It's then up to the user to say "I promise I know what I'm doing" by unsafely implementing `FinalizerSafe` on the type passed to `Gc::new`.

ltratt replied:
> So all we're doing here is easing up a bit and letting more (but obviously not all) correct programs through.

Ah, so we've got a safe underapproximation of correct programs? If so, I'm happy with this PR as-is, and the next PR can improve the underapproximation. But if you want to do that in this one PR, that's also OK with me!

ltratt commented on compiler/rustc_mir_transform/src/check_finalizers.rs:
```diff
@@ -70,18 +48,186 @@ impl<'tcx> MirPass<'tcx> for CheckFinalizers {
+                        .unwrap();
+                    let mir = self.tcx.instance_mir(instance.def);
+                    let mut checker = ProjectionChecker { cx: self, body: mir };
+                    checker.visit_body(&mir);
```
I'm not familiar with rustc's innards: is this the bit which (indirectly) does the "call `visit_projection`"?

jacob-hughes replied:
That's right, this kicks off the `MirVisitor` which walks the CFG and then executes our code when it sees a projection.

ltratt replied:
Got it!

jacob-hughes commented:
I've fixed the bug which allowed making calls with `self` args inside drop here 710118feec141082dfe31af1a9ec663d03a225ef. I've also pushed some formatting changes and implemented `FinalizerSafe` on every standard library type where it's appropriate. This PR is now ready for re-review.

ltratt commented on compiler/rustc_mir_transform/src/check_finalizers.rs:
````diff
@@ -189,44 +212,60 @@ impl<'a, 'tcx> Visitor<'tcx> for ProjectionChecker<'a, 'tcx> {
+                        // like below from compiling:
+                        //
+                        // ```
+                        // fn drop(&mut self) {
````
We should add this example to the test suite and add a similar comment along the lines of "this program should work but our underapproximation causes it not to". That way, if we ever fiddle with this code, we'll know if we've fixed or broken programs we thought should/shouldn't work!

ltratt commented:
Please squash.

jacob-hughes commented:
Squashed

ltratt commented:
bors r+

jacob-hughes commented:
Squashed

ltratt commented:
bors r+

jacob-hughes commented:
bors try

ltratt commented:
If `bors` passes, then please squash. If not... then don't :)

jacob-hughes commented:
Thanks. I still can't reproduce this locally. It's fiddly because `backtrace_rs` pulls in a different frame layout depending on the target.

jacob-hughes commented:
bors try

jacob-hughes commented:
Finally! Squashed
